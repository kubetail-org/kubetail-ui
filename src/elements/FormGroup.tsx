// Copyright 2024 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createContext, forwardRef, useMemo } from 'react';

import { cn } from '@/lib/utils';

type ContextState = {
  controlId: string | undefined,
};

const Context = createContext({} as ContextState); // for passing `controlId`
const baseCls = 'space-y-1';

interface Props extends React.ComponentPropsWithoutRef<'div'> {
  controlId?: string;
}

const FormGroup = forwardRef((
  {
    className,
    controlId,
    ...props
  }: Props,
  ref: React.ForwardedRef<HTMLDivElement>,
) => {
  const context = useMemo(() => ({ controlId }), [controlId]);

  return (
    <Context.Provider value={context}>
      <div
        {...props}
        ref={ref}
        className={cn(baseCls, className)}
      />
    </Context.Provider>
  );
});

FormGroup.displayName = 'FormGroup';

export { Context };
export default FormGroup;
