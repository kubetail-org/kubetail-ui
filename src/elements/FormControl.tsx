// Copyright 2024-2025 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { forwardRef, useContext } from 'react';

import { cn } from '@/lib/utils';

import { Context as FormGroupContext } from './FormGroup';
import FormControlFeedback from './FormControlFeedback';

const baseCls =
  'flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium';

interface Props extends React.ComponentPropsWithoutRef<'input'> {
  as?: React.ElementType;
  id?: string;
}

const FormControl = forwardRef(
  ({ as = 'input', className, id, ...props }: Props, ref: React.ForwardedRef<HTMLInputElement>) => {
    const Tag = as;
    const { controlId } = useContext(FormGroupContext);

    return <Tag {...props} ref={ref} id={id || controlId} className={cn(baseCls, className)} />;
  },
);

FormControl.displayName = 'FormControl';

type FormControlType = typeof FormControl;

interface FormControlExportType extends FormControlType {
  Feedback: typeof FormControlFeedback;
}

const FormControlExport: FormControlExportType = Object.assign(FormControl, {
  Feedback: FormControlFeedback,
});

export default FormControlExport;
