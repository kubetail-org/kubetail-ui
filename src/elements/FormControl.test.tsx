// Copyright 2024-2025 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { render } from '@testing-library/react';

import FormControl from './FormControl';
import FormGroup from './FormGroup';

describe('FormControl', () => {
  it('renders properly', () => {
    const { asFragment } = render(<FormControl type="text" defaultValue="content" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('appends custom className', () => {
    const { asFragment } = render(<FormControl className="extra-class" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders `disabled` properly', () => {
    const { asFragment } = render(<FormControl disabled />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders `disabled=false` properly', () => {
    const { asFragment } = render(<FormControl disabled={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as textarea properly', () => {
    const { asFragment } = render(<FormControl as="textarea" defaultValue="contents" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('inherits id from FormGroup.controlId', () => {
    const { asFragment } = render(
      <FormGroup controlId="input-id">
        <FormControl />
      </FormGroup>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
