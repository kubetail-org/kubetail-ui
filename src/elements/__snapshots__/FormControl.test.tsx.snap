// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormControl > appends custom className 1`] = `
<DocumentFragment>
  <input
    class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium extra-class"
  />
</DocumentFragment>
`;

exports[`FormControl > inherits id from FormGroup.controlId 1`] = `
<DocumentFragment>
  <div
    class="space-y-1"
  >
    <input
      class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium"
      id="input-id"
    />
  </div>
</DocumentFragment>
`;

exports[`FormControl > renders \`disabled=false\` properly 1`] = `
<DocumentFragment>
  <input
    class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium"
  />
</DocumentFragment>
`;

exports[`FormControl > renders \`disabled\` properly 1`] = `
<DocumentFragment>
  <input
    class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium"
    disabled=""
  />
</DocumentFragment>
`;

exports[`FormControl > renders as textarea properly 1`] = `
<DocumentFragment>
  <textarea
    class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium"
  >
    contents
  </textarea>
</DocumentFragment>
`;

exports[`FormControl > renders properly 1`] = `
<DocumentFragment>
  <input
    class="flex h-10 w-full rounded border border-input bg-background px-3 py-2 text-sm placeholder:text-muted-foreground ring-offset-background focus:outline-none focus:border-input focus:ring focus:ring-ring focus:ring-offset-0 disabled:cursor-not-allowed disabled:opacity-50 file:border-0 file:bg-transparent file:text-sm file:font-medium"
    type="text"
    value="content"
  />
</DocumentFragment>
`;
